fe/fe-core/src/main/java/org/apache/doris/qe/PointQueryExecutor.java

```diff
@@ -0,0 +1,351 @@
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.thrift.TDeserializer;
+import org.apache.thrift.TException;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class PointQueryExecutor implements CoordInterface {
+    private static final Logger LOG = LogManager.getLogger(PointQueryExecutor.class);
+    private long tabletID = 0;
+    private long timeoutMs = Config.point_query_timeout_ms; // default 10s
+
+    private boolean isCancel = false;
+    private List<Backend> candidateBackends;
+    private final int maxMsgSizeOfResultReceiver;
+
+    // used for snapshot read in cloud mode
+    private List<Long> versions;
+
+    private final ShortCircuitQueryContext shortCircuitQueryContext;
+
+    public PointQueryExecutor(ShortCircuitQueryContext ctx, int maxMessageSize) {
+        ctx.sanitize();
+        this.shortCircuitQueryContext = ctx;
+        this.maxMsgSizeOfResultReceiver = maxMessageSize;
+    }
+
+    private void updateCloudPartitionVersions() throws RpcException {
+        OlapScanNode planRoot = shortCircuitQueryContext.scanNode;
+        List<CloudPartition> partitions = new ArrayList<>();
+        Set<Long> partitionSet = new HashSet<>();
+        OlapTable table = planRoot.getOlapTable();
+        for (Long id : planRoot.getSelectedPartitionIds()) {
+            if (!partitionSet.contains(id)) {
+                partitionSet.add(id);
+                partitions.add((CloudPartition) table.getPartition(id));
+            }
+        }
+        versions = CloudPartition.getSnapshotVisibleVersion(partitions);
+        // Only support single partition at present
+        Preconditions.checkState(versions.size() == 1);
+        LOG.debug("set cloud version {}", versions.get(0));
+    }
+
+    void setScanRangeLocations() throws Exception {
+        OlapScanNode planRoot = shortCircuitQueryContext.scanNode;
+        // compute scan range
+        List<TScanRangeLocations> locations = planRoot.lazyEvaluateRangeLocations();
+        if (planRoot.getScanTabletIds().isEmpty()) {
+            return;
+        }
+        Preconditions.checkState(planRoot.getScanTabletIds().size() == 1);
+        this.tabletID = planRoot.getScanTabletIds().get(0);
+
+        // update partition version if cloud mode
+        if (Config.isCloudMode()
+                && ConnectContext.get().getSessionVariable().enableSnapshotPointQuery) {
+            // TODO: Optimize to reduce the frequency of version checks in the meta service.
+            updateCloudPartitionVersions();
+        }
+
+        Preconditions.checkNotNull(locations);
```

Comment:
Is locations only used by checkNotNull? If yes, move it just after locations = xxx for more readable.